internal/utils/utils.go

```diff
@@ -1,56 +1,179 @@
 package utils
 
 import (
+	"fmt"
 	"io"
 	"net/http"
 	"strconv"
+	"strings"
+	"sync"
 
 	log "github.com/sirupsen/logrus"
 )
 
-type NetUtil struct {
-	BufferSize int
-}
+var (
+	bufferSize int
+	debug      bool
+)
 
 type writerWrapper struct {
 	writer    io.Writer
 	writeable bool
 }
 
-func (netutil NetUtil) Forward(src *http.Response, dst http.ResponseWriter) error {
+// Initialize .
+func Initialize(_bufferSize int, _debug bool) {
+	bufferSize = _bufferSize
+	debug = _debug
+}
+
+// Forward .
+func Forward(src *http.Response, dst http.ResponseWriter) error {
 	CopyHeader(src, dst)
-	_, err := netutil.readAndWriteToDst(src.Body, dst, true)
-	return err
+	return copyBody(src.Body, dst)
+}
+
+// Link .
+func Link(client io.ReadWriteCloser, server io.ReadWriteCloser) {
+	wg := sync.WaitGroup{}
+	wg.Add(2)
+	go forwardByteStream(client, server, &wg, "from client")
+	go forwardByteStream(server, client, &wg, "from server")
+	wg.Wait()
+}
+
+func forwardByteStream(src io.Reader, dst io.WriteCloser, wg *sync.WaitGroup, label string) {
+	defer wg.Done()
+	defer dst.Close()
+
+	log.Print("Starting forward byte stream ", label)
+	if _, err := io.Copy(dst, src); err != nil {
```

Comment:
is it may blocked?

Reply:
yes, but the block doesn't matter, for now on it is tcp streaming